src/engine.rs

```diff
@@ -288,6 +344,58 @@ impl Engine {
         sdk::sha256(&data)
     }
 
+    pub fn charge_gas_limit(
+        sender: &Address,
+        gas_limit: U256,
+        gas_price: U256,
+    ) -> Result<Wei, GasPaymentError> {
+        // Early exit as performance optimization
+        if gas_price.is_zero() {
+            return Ok(Wei::zero());
+        }
+
+        let eth_amount = Wei::new(
+            gas_limit
+                .checked_mul(gas_price)
+                .ok_or(GasPaymentError::EthAmountOverflow)?,
+        );
+        let account_balance = Self::get_balance(sender);
+        let remaining_balance = account_balance
+            .checked_sub(eth_amount)
+            .ok_or(GasPaymentError::OutOfFund)?;
+
+        Self::set_balance(sender, &remaining_balance);
+
+        Ok(eth_amount)
+    }
+
+    pub fn refund_unused_gas(
+        sender: &Address,
+        relayer: &Address,
+        prepaid_eth: Wei,
+        used_gas: u64,
+        gas_price: U256,
+    ) -> Result<(), GasPaymentError> {
+        // Early exit as performance optimization
+        if gas_price.is_zero() {
+            return Ok(());
+        }
+
+        let used_amount = Wei::new(
+            gas_price
+                .checked_mul(used_gas.into())
+                .ok_or(GasPaymentError::EthAmountOverflow)?,
+        );
+        // We cannot have used more than the gas_limit
+        debug_assert!(used_amount <= prepaid_eth);
+        let refund_amount = prepaid_eth - used_amount;
```

Comment:
```suggestion
        debug_assert!(used_amount <= prepaid_amount_for_gas);
        let refund_amount = prepaid_amount_for_gas - used_amount;
```

Reply:
Done in https://github.com/aurora-is-near/aurora-engine/pull/215/commits/5bdc6d0dc4897e5892f0f72d8e1eee36602a45e0